core/types/system_tx.go

```diff
@@ -0,0 +1,109 @@
+package types
+
+import (
+	"bytes"
+	"math/big"
+	"sort"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/rlp"
+)
+
+type SystemTx struct {
+	From  common.Address  // pre-determined sender
+	To    *common.Address // system contract
+	Gas   uint64          // gas limit
+	Nonce uint64          // nonce
+	Data  []byte          // calldata
+}
+
+func (tx *SystemTx) txType() byte { return SystemTxType }
+
+func (tx *SystemTx) copy() TxData {
+	return &SystemTx{
+		From:  tx.From,
+		Gas:   tx.Gas,
+		Nonce: tx.Nonce,
+		To:    copyAddressPtr(tx.To),
+		Data:  common.CopyBytes(tx.Data),
+	}
+}
+
+func (tx *SystemTx) chainID() *big.Int      { return new(big.Int) }
+func (tx *SystemTx) accessList() AccessList { return nil }
+func (tx *SystemTx) data() []byte           { return tx.Data }
+func (tx *SystemTx) gas() uint64            { return tx.Gas }
+func (tx *SystemTx) gasPrice() *big.Int     { return new(big.Int) }
+func (tx *SystemTx) gasTipCap() *big.Int    { return new(big.Int) }
+func (tx *SystemTx) gasFeeCap() *big.Int    { return new(big.Int) }
+func (tx *SystemTx) value() *big.Int        { return new(big.Int) }
+func (tx *SystemTx) nonce() uint64          { return tx.Nonce }
+func (tx *SystemTx) to() *common.Address    { return tx.To }
+
+func (tx *SystemTx) rawSignatureValues() (v, r, s *big.Int) {
+	return new(big.Int), new(big.Int), new(big.Int)
+}
+
+func (tx *SystemTx) setSignatureValues(chainID, v, r, s *big.Int) {}
+
+func (tx *SystemTx) encode(b *bytes.Buffer) error {
+	return rlp.Encode(b, tx)
+}
+
+func (tx *SystemTx) decode(input []byte) error {
+	return rlp.DecodeBytes(input, tx)
+}
+
+var _ TxData = (*SystemTx)(nil)
+
+func NewOrderedSystemTxs(stxs []*SystemTx) *OrderedSystemTxs {
```

Comment:
I'm not sure why we need this. The system txs should be generated by the worker sequentially. There is no need to re-order them.

Reply:
for one sender there is no diff.
We do need to guarantee order for Pop to work correctly however.

Maybe we can make that a noop, and then we can avoid the sort.

Reply:
`worker` uses `OrderedTransactionSet` (which `OrderedSystemTxs` implements) to decide the order in which it processes transactions in `commitTransactions`. It was mostly added for compatibility with geth's original logic. But yeah, for `SystemTx` we simply keep the original order.

Reply:
fixed, `Pop` is a noop, and there is no sorting.

Reply:
Yes, we can just use `[]types.SystemTx` in the worker similar to [here](https://github.com/scroll-tech/go-ethereum/blob/733dd01a9af05c78f617731127a1aeae93a6cd0b/miner/worker.go#L1326). We can remove the `OrderedSystemTxs` struct.

Reply:
we produce that in emitSystemTxs, but we need the order set for processing them.